### PR TITLE
feature: Parameterized queries and transaction API

### LINKED ISSUE #17: Parameterized queries
In order to support parameterized queries we need to use the extended postgres wire protocol to pass queries.

### LINKED ISSUE #29: Weird inconsistency with real postgres
I have a library that generates SQL that looks like this
```sql
with "cte1" as (
  (select arg_0 as "a", arg_1 as "b" from (values (1, 'Foo'), (2, 'Bar'), (3, 'Baz')) as vals(arg_0, arg_1))
)
select "a", "b" from "cte1";
```

In real postgres it returns correctly
```
psql (16.2 (Debian 16.2-1.pgdg120+2))
Type "help" for help.

postgres=# with "cte1" as (
  (select arg_0 as "a", arg_1 as "b" from (values (1, 'Foo'), (2, 'Bar'), (3, 'Baz')) as vals(arg_0, arg_1))
)
select "a", "b" from "cte1";
 a |  b
---+-----
 1 | Foo
 2 | Bar
 3 | Baz
(3 rows)
```

In pglite I get weird answers
```js
import { PGlite } from "@electric-sql/pglite"

const db = new PGlite();

const res = await db.query(`
  with "cte1" as (
    (select arg_0 as "a", arg_1 as "b" from (values (1, 'Foo'), (2, 'Bar'), (3, 'Baz')) as vals(arg_0, arg_1))
  )
  select "a", "b" from "cte1";
`);
console.log(res);
```
```shell
> bun foo.mjs
[
  {
    "": "Foo",
  }, {
    "": "Bar",
  }, {
    "": "Baz",
  }
]
```

```shell
> deno run -A foo.mjs
[ { "": "Foo" }, { "": "Bar" }, { "": "Baz" } ]
```

### LINKED ISSUE #31: Use the Postgres wire protocol for all input+output

### PR DESCRIPTION
**Pull request by samwillis**
Each PGlite instance (in that PR) has two methods to run a query:
- `.query(sql, params)`: execute a single statement, optionally with params. Returns single result object
- `.exec(sql)`: execute one or more statements, no params. Returns array of result objects, one for each statement.

I split this into two as it makes it easer to know if you are getting one or multiple result objects. Internally the first uses the extended protocol which only supports a single statement, the latter the simple protocol that can run multiple statements but with no params.

Node: this is a breaking change, what was `.query(sql)` is now `.exec(sql)`.

It also has a `.transaction((tx) => {})` api for interactive transactions. Each `tx` object also has the same `.query` and `.exec` methods, along with a `.rollback()`.

Finally there is also a `.execProtocol(msg: uInt8Array)` method that allows users to execute raw pg-protocl message.

If anyone wants to try this, there is a build including this linked from: #48 

Fixes #17, Fixes #31, Fixes #29, see also: https://github.com/electric-sql/postgres-wasm/pull/3, https://github.com/electric-sql/postgres-wasm/pull/4

### CODE REVIEW

**Review by msfstef — COMMENTED** [review]
Amazing work! I think it's quite clean and easy to follow, I've left a few nits and comments mostly focusing on clarity.

I think the only design decision I'm not sure about is the transaction rollback API. It might be worth getting rid of it and just letting any error result in a rollback and needing the user to handle it. I think no transactions should aim to fail and be rolled back, so any rollbacks should be viewed as the internal preparation for the error handling logic left to the user.

File: packages/pglite/package.json [hunk 1]
```diff
@@ -17,11 +24,11 @@
   ],
   "repository": {
     "type": "git",
-    "url": "https://github.com/electric-sql/",
+    "url": "https://github.com/electric-sql/pglite",
     "directory": "packages/pglite"
   },
   "scripts": {
-    "test": "ava",
+    "test": "rm -rf ./pgdata-test && concurrently --hide 0 --prefix none -k \"npx http-server --port 3334 ./\" \"sleep 2 && ava tests/*.test.js tests/**/*.test.js\"",
```
[the remark this answers is not in the document]
Comment (samwillis) [reply packages/pglite/package.json:31]: Yep, the http-server shuts down after the tests.

I've loged an issue for Windows build support: #43

**Comment by samwillis** [comment]
Thanks @msfstef!

I've addressed everything in your review.

For the transactions, `.rollback()` is now an async call that issues the rollback and then marks the transaction as closed so no further operations can be applied. The user can then continue with their own cleanup and return whatever they want from the transaction call. Could take a quick look at that and see if you think it works well?

**Comment by msfstef** [comment]
@samwillis I like the idea of allowing the user to perform their cleanup within the transaction - though the cleanup will still hold the lock without strictly needing it. Might be still worth documenting that transactions get rolled back on _any_ error though regardless of whether `rollback()` is called.

Looks good, ready to merge!

**Review by msfstef — APPROVED** [review]

**Comment by AntonOfTheWoods** [comment]
Is there a build of this with the changes somewhere? It would be nice to take it out for a spin!

**Comment by samwillis** [comment]
Hey @AntonOfTheWoods (and anyone else),

There is a dev build of the latest changes from the tip of my stacked PRs linked on #48. You can download and try it out.

We are planning to do a release early next week after a few final checks.

**Pull request merged**